# 4.8 Grouping Data & Aggregating Variables
### TABLE OF CONTENTS
#### Importing libraries
#### Grouping data with pandas
#### Aggregating Data with agg()
#### Aggregating Data with transform()
#### Deriving Columns with loc()
#### EXERCISE STEP 1: Setup notebook
#### STEP 2: Repeat aggregated mean for the entire dataframe
#### STEP 3: Analyze the result
#### STEP 4: Create loyalty flag for existing customers using transform() and loc()
#### STEP 5: User loyalty flag to check product price statistics for each loyalty category
#### STEP 6: Create a spending flag
#### STEP 7: Create a frequency flag
#### STEP 8: Ensure notebook is clean and structured
#### STEP 9: Export dataframe as a pickle file

#01. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# Creating path

path = r'/Users/kristincpeterson/Desktop/DataAnalytics/Achievement_4/Instacart Basket Analysis'

In [3]:
path

'/Users/kristincpeterson/Desktop/DataAnalytics/Achievement_4/Instacart Basket Analysis'

In [4]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02_Data', 'Prepared_Data', 'orders_products_merged_updated_extra_columns.pkl'))

In [5]:
ords_prods_merge.shape

(32434212, 17)

In [6]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_day_of_week  \
0   3139998      138    prior            28                  6   
1   1977647      138    prior            30                  6   
2    389851      709    prior             2                  0   
3    652770      764    prior             1                  3   
4   1813452      764    prior             3                  4   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 13                     NaN                 10          0   
4                 17                     9.0                 11          1   

      busiest_day         busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy days           Most orders  
1  Regularly busy  Regularly busy days        Average orders  
2     Busiest day         Busiest days        Average orders  
3  Regularly busy      Least busy days           Most orders  
4      Least busy      Least busy days        Average orders

In [7]:
# Create a subset

df = ords_prods_merge [:1000000]

In [8]:
df.shape

(1000000, 17)

In [9]:
df.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_day_of_week  \
0   3139998      138    prior            28                  6   
1   1977647      138    prior            30                  6   
2    389851      709    prior             2                  0   
3    652770      764    prior             1                  3   
4   1813452      764    prior             3                  4   
5   1701441      777    prior            16                  1   
6   1871483      825    prior             3                  2   
7   1290456      910    prior            12                  3   
8    369558     1052    prior            10                  1   
9    589712     1052    prior            15                  1   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 13                     NaN                 10          0   
4                 17                     9.0                 11          1   
5                  7                    26.0                  7          0   
6                 14                    30.0                  2          0   
7                 10                    30.0                  1          0   
8                 20                    19.0                  1          0   
9                 12                    15.0                  2          1   

      busiest_day         busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy days           Most orders  
1  Regularly busy  Regularly busy days        Average orders  
2     Busiest day         Busiest days        Average orders  
3  Regularly busy      Least busy days           Most orders  
4      Least busy      Least busy days        Average orders  
5  Regularly busy         Busiest days        Average orders  
6  Regularly busy  Regularly busy days           Most orders  
7  Regularly busy      Least busy days           Most orders  
8  Regularly busy         Busiest days        Average orders  
9  Regularly busy         Busiest days           Most orders

## Grouping Data with pandas

In [10]:
df.groupby('product_name')

## Aggregating Data with agg()

In [11]:
#Performing a single aggregation

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.562844
2                17.518895
3                16.907152
4                17.537934
5                14.894521
6                16.658449
7                17.072954
8                14.942021
9                15.447580
10               18.681852
11               15.352278
12               14.327957
13               16.578690
14               16.666369
15               16.153766
16               17.768070
17               15.602964
18               19.674252
19               16.924344
20               16.292631
21               25.535596

In [12]:
# Some aggregations can be conducted without use of the agg() function. For instance:

df.groupby('department_id')['order_number'].mean()

department_id
1     15.562844
2     17.518895
3     16.907152
4     17.537934
5     14.894521
6     16.658449
7     17.072954
8     14.942021
9     15.447580
10    18.681852
11    15.352278
12    14.327957
13    16.578690
14    16.666369
15    16.153766
16    17.768070
17    15.602964
18    19.674252
19    16.924344
20    16.292631
21    25.535596
Name: order_number, dtype: float64

In [13]:
# Rather than including the column name within square brackets, as in the example above, you could also set the column name off with a dot, for instance:

df.groupby('department_id').order_number.mean()

department_id
1     15.562844
2     17.518895
3     16.907152
4     17.537934
5     14.894521
6     16.658449
7     17.072954
8     14.942021
9     15.447580
10    18.681852
11    15.352278
12    14.327957
13    16.578690
14    16.666369
15    16.153766
16    17.768070
17    15.602964
18    19.674252
19    16.924344
20    16.292631
21    25.535596
Name: order_number, dtype: float64

### Performing Multiple Aggregations

In [14]:
# Producing multiple statistics at the same time

df.groupby('department_id').agg({'order_number' : ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.562844   1  99
2                17.518895   1  96
3                16.907152   1  99
4                17.537934   1  99
5                14.894521   1  99
6                16.658449   1  99
7                17.072954   1  99
8                14.942021   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.352278   1  99
12               14.327957   1  99
13               16.578690   1  99
14               16.666369   1  99
15               16.153766   1  99
16               17.768070   1  99
17               15.602964   1  99
18               19.674252   1  99
19               16.924344   1  99
20               16.292631   1  99
21               25.535596   1  98

## Aggregating Data with transform()

In [15]:
# Create a loyalty flag -- 3-step process in one line of code:
# 01. Split the data into groups based on the "user_id" column
# 02. Apply the transform() function on the "order_number" column to generate the maximum orders for each user
# 03. Create a new column, "max_order", into which you'll place the results of your new aggregation

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [16]:
# Checking results

ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  order_day_of_week  \
0    3139998      138    prior            28                  6   
1    1977647      138    prior            30                  6   
2     389851      709    prior             2                  0   
3     652770      764    prior             1                  3   
4    1813452      764    prior             3                  4   
5    1701441      777    prior            16                  1   
6    1871483      825    prior             3                  2   
7    1290456      910    prior            12                  3   
8     369558     1052    prior            10                  1   
9     589712     1052    prior            15                  1   
10   1798115     1379    prior             1                  5   
11    497193     1480    prior            36                  3   
12   1727826     1480    prior            38                  4   
13   2694698     1480    prior            39                  2   
14   3171753     1494    prior             6                  3   

    order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                  11                     3.0                  5          0   
1                  17                    20.0                  1          1   
2                  21                     6.0                 20          0   
3                  13                     NaN                 10          0   
4                  17                     9.0                 11          1   
5                   7                    26.0                  7          0   
6                  14                    30.0                  2          0   
7                  10                    30.0                  1          0   
8                  20                    19.0                  1          0   
9                  12                    15.0                  2          1   
10                 11                     NaN                  1          0   
11                 13                     7.0                  1          0   
12                 13                     7.0                  3          1   
13                 12                    12.0                  4          1   
14                 10                     2.0                  8          0   

       busiest_day         busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy days           Most orders         32  
1   Regularly busy  Regularly busy days        Average orders         32  
2      Busiest day         Busiest days        Average orders          5  
3   Regularly busy      Least busy days           Most orders          3  
4       Least busy      Le

In [17]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
..         ...                         ...       ...            ...     ...   
95           1  Chocolate Sandwich Cookies        61             19     5.8   
96           1  Chocolate Sandwich Cookies        61             19     5.8   
97           1  Chocolate Sandwich Cookies        61             19     5.8   
98           1  Chocolate Sandwich Cookies        61             19     5.8   
99           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  order_day_of_week  \
0    3139998      138    prior            28                  6   
1    1977647      138    prior            30                  6   
2     389851      709    prior             2                  0   
3     652770      764    prior             1                  3   
4    1813452      764    prior             3                  4   
..       ...      ...      ...           ...                ...   
95    602103    10831    prior             8                  3   
96     49629    11119    prior            18                  1   
97    317888    11186    prior            13                  5   
98    682486    11243    prior            16                  3   
99   2861386    11448    prior            10                  2   

    order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                  11                     3.0                  5          0   
1                  17                    20.0                  1          1   
2                  21                     6.0                 20          0   
3                  13                     NaN                 10          0   
4                  17                     9.0                 11          1   
..                ...                     ...                ...        ...   
95                 11                    23.0                  5          0   
96                 14                    30.0                  1          0   
97                 16                     2.0                  8          0   
98                 13                     0.0                  2          0   
99                 15                     3.0                  6          0   

       busiest_day         busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy days           Most orders         32  
1   Regularly busy  Regularly busy days        Average orders         32  
2      Busiest day         Busiest days        Average orders          5  
3   Regularly busy      Least busy days           Most orders          3  
4       Least busy      Least busy days        Average orders          3  
..             ...                  ...                   ...        ...  
95  Regularly busy      Least busy days           Most orders         10  
96  Regularly busy         Busiest days           Most orders         23  
97  Regularly busy  Regularly busy days           Most orders         26  
98  Regularly busy      Least busy days           Most orders         43  
99  Regularly busy  Regularly busy days           Most orders         21  

[100 rows x 18 columns]

In [18]:
# Change display to show all of the rows

pd.options.display.max_rows = None

In [19]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

## Deriving Columns with loc()

In [20]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [21]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [22]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [23]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [24]:
# Checking results

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

# Exercise 4.8

## STEP 1: Done

## STEP 2: Find the aggregated mean of the "order_number" column grouped by "department_id" for the entire dataframe.

In [25]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## STEP 3: Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

In [26]:
# Run subset code again for comparison

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.562844
2                17.518895
3                16.907152
4                17.537934
5                14.894521
6                16.658449
7                17.072954
8                14.942021
9                15.447580
10               18.681852
11               15.352278
12               14.327957
13               16.578690
14               16.666369
15               16.153766
16               17.768070
17               15.602964
18               19.674252
19               16.924344
20               16.292631
21               25.535596

#### While not drastically different numbers, the subset averages tend to be slightly higher overall. The subset min is 14.3 (vs min of 15.2 for the entire data set), and the subset max is 25.5 (vs max of 22.9 on entire data set).

## STEP 4: Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [27]:
# Step 1 - Split the data into groups based on the “user_id” column.

ords_prods_merge.groupby('user_id')

In [28]:
# Step 2 - Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
# Step 3 - Create a new column, “max_order,” into which you’ll place the results of your aggregation.

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [29]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  order_day_of_week  \
0    3139998      138    prior            28                  6   
1    1977647      138    prior            30                  6   
2     389851      709    prior             2                  0   
3     652770      764    prior             1                  3   
4    1813452      764    prior             3                  4   
5    1701441      777    prior            16                  1   
6    1871483      825    prior             3                  2   
7    1290456      910    prior            12                  3   
8     369558     1052    prior            10                  1   
9     589712     1052    prior            15                  1   
10   1798115     1379    prior             1                  5   
11    497193     1480    prior            36                  3   
12   1727826     1480    prior            38                  4   
13   2694698     1480    prior            39                  2   
14   3171753     1494    prior             6                  3   

    order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                  11                     3.0                  5          0   
1                  17                    20.0                  1          1   
2                  21                     6.0                 20          0   
3                  13                     NaN                 10          0   
4                  17                     9.0                 11          1   
5                   7                    26.0                  7          0   
6                  14                    30.0                  2          0   
7                  10                    30.0                  1          0   
8                  20                    19.0                  1          0   
9                  12                    15.0                  2          1   
10                 11                     NaN                  1          0   
11                 13                     7.0                  1          0   
12                 13                     7.0                  3          1   
13                 12                    12.0                  4          1   
14                 10                     2.0                  8          0   

       busiest_day         busiest_days busiest_period_of_day  max_order  \
0   Regularly busy  Regularly busy days           Most orders         32   
1   Regularly busy  Regularly busy days        Average orders         32   
2      Busiest day         Busiest days        Average orders          5   
3   Regularly busy      Least busy days           Most orders          3   
4       Least busy   

## STEP 5: The marketing team at Instacart wants to know whether there's a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you're trying to determine is whether the prices of products purchased by loyalty customers differ from those purchased by regular or new customers.

In [30]:
# Group by "prices" 

ords_prods_merge.groupby('prices')

In [31]:
# Split the data into groups by "loyalty_flag" and then calculate the mean for each "prices" column

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.388747
New customer      13.294370
Regular customer  12.496203

## STEP 6: The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders.

In [32]:
# Group data by "user_id" column, generate mean product price for each user, create "mean_price" for aggregation results

ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [33]:
# Check grouping

ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [34]:
# Create a spending flag if the mean prices of products purchased by a user is less than or equal to 10, then flag them as a "Low spender"

ords_prods_merge.loc[ords_prods_merge['mean_price'] <=10, 'spending_flag'] = 'Low spender'

In [35]:
# Create a spending flag if the mean prices of products purchased by a user is higher than or equal to 10, then flag them as a "High spender"

ords_prods_merge.loc[ords_prods_merge['mean_price'] >=10, 'spending_flag'] = 'High spender'

In [36]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [37]:
ords_prods_merge[['user_id', 'mean_price', 'loyalty_flag']].head(60)

user_id  mean_price      loyalty_flag
0       138    6.935811  Regular customer
1       138    6.935811  Regular customer
2       709    7.930208      New customer
3       764    4.972414      New customer
4       764    4.972414      New customer
5       777    6.935398  Regular customer
6       825    5.957576      New customer
7       910    6.680000  Regular customer
8      1052    7.162500  Regular customer
9      1052    7.162500  Regular customer
10     1379    5.813636      New customer
11     1480    5.139394    Loyal customer
12     1480    5.139394    Loyal customer
13     1480    5.139394    Loyal customer
14     1494    6.803797  Regular customer
15     1494    6.803797  Regular customer
16     1494    6.803797  Regular customer
17     1540    6.607122  Regular customer
18     1540    6.607122  Regular customer
19     1540    6.607122  Regular customer
20     1540    6.607122  Regular customer
21     1540    6.607122  Regular customer
22     1540    6.607122  Regular customer
23     1540    6.607122  Regular customer
24     1540    6.607122  Regular customer
25     1540    6.607122  Regular customer
26     1540    6.607122  Regular customer
27     1540    6.607122  Regular customer
28     1540    6.607122  Regular customer
29     1540    6.607122  Regular customer
30     1540    6.607122  Regular customer
31     1540    6.607122  Regular customer
32     1540    6.607122  Regular customer
33     1540    6.607122  Regular customer
34     1598    7.123529      New customer
35     1598    7.123529      New customer
36     1598    7.123529      New customer
37     1647    4.853521      New customer
38     1647    4.853521      New customer
39     1647    4.853521      New customer
40     1647    4.853521      New customer
41     1647    4.853521      New customer
42     2833    7.377143      New customer
43     2850    8.054153  Regular customer
44     2857    8.387075  Regular customer
45     3010    6.452800  Regular customer
46     3029    7.960606      New customer
47     3395    6.322222      New customer
48     3904    7.511111      New customer
49     4006    7.441100    Loyal customer
50     4006    7.441100    Loyal customer
51     4006    7.441100    Loyal customer
52     4006    7.441100    Loyal customer
53     4006    7.441100    Loyal customer
54     4122    7.598765  Regular customer
55     4141    6.033333      New customer
56     4343    7.393939      New customer
57     4553    5.715789      New customer
58     4845    7.229630  Regular customer
59     6866    4.807143      New customer

## STEP 7: In order to send relevant notifications to users wtihin the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user's ordering behavior according to the median in the "days_since_prior_order" column.

In [38]:
# Group data by "user_id" column, generate median product of "days since prior order, create "median_days" for aggregation results

ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [39]:
# Check grouping

ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [40]:
# Create flags to assign "freqency" labels to user ID based on median days since prior order

ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [41]:
ords_prods_merge.loc[(ords_prods_merge['median_days'] > 10) & (ords_prods_merge['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [42]:
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [43]:
# Check results

ords_prods_merge['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent customer        21577409
Regular customer          7217134
Non-frequent customer     3639669
Name: count, dtype: int64

## STEP 8: Ensure notebook is clean and structured

#### Done.

In [44]:
# Exporting new df

ords_prods_merge.to_pickle(os.path.join(path, '02_Data', 'Prepared_Data', 'orders_products_newest.pkl'))

In [45]:
# Checking for missing values

cust.isnull().sum()

NameError: name 'cust' is not defined